circuits/src/permutations.nr

```diff
@@ -18,6 +18,72 @@ fn absorb(input: [u1; BLOCK_SIZE]) -> [u1; STATE_SIZE] {
   };
 
   // TODO: apply keccak-p functions
+  state = keccakf(state);
+  
+  state
+}
+
+fn squeeze(input: [u1; STATE_SIZE]) -> [u1; 256] {
+  
+  // TODO: after PR for allowing expressions to be evaluated inside arrays gets merged in change array literal to [0 as u1; num_bits_out]
+  let mut result: [u1; 256] = [0 as u1; 256];
+
+  for i in 0..STATE_SIZE {
+    if (i < 256) {
+      result[i] = input[i];
+    }
+  };
+  result
+}
+
+fn keccakfRound(state: [u1; STATE_SIZE], round_number: comptime Field) -> [u1; STATE_SIZE] {
+
+  let state_after_theta = theta::theta(state);
+  let state_after_rhoPi = rhoPi::rhoPi(state_after_theta);
+  let state_after_chi = chi::chi(state_after_rhoPi);
+  let new_state = iota::iota(state_after_chi, round_number);
+
+  new_state
+}
+
+fn keccakf(input: [u1; STATE_SIZE]) -> [u1; STATE_SIZE] {
+  let mut state: [u1; STATE_SIZE] = [0 as u1; 1600];
+  let mut mid_round: [u1; 38400] = [0 as u1; 38400];
+  for i in 0..24 {
+    if (i == 0) {
+        for j in 0..STATE_SIZE {
+          mid_round[j] = input[j];
+        }
+    }
 
+    let mut temp_state: [u1; STATE_SIZE] = [0 as u1; 1600];
+    for j in 0..STATE_SIZE {
+      temp_state[j] = mid_round[(i*STATE_SIZE)+j];
+    };
+    let round_result = keccakfRound(temp_state, i);
+    if (i < 23) {
+      for j in 0..STATE_SIZE {
+        mid_round[((i+1)*STATE_SIZE)+j] = round_result[j];
+      };
+    }
+    state = round_result;
+  };
```

Comment:
I'm interested in the role of `mid_round` here. Can we not just overwrite the state vector here rather than copying out the state 24 times?

Reply:
I think you are correct here. I've changed it and it is much cleaner